File: tests/wasm/test_ehlers_simple_cycle_indicator.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { loadTestData, assertArrayClose } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

function manualReference(data, alpha = 0.07) {
    const cycle = new Float64Array(data.length);
    const trigger = new Float64Array(data.length);
    const smooth = new Float64Array(data.length);
    cycle.fill(NaN);
    trigger.fill(NaN);
    smooth.fill(NaN);

    const coeffCycle = (1 - 0.5 * alpha) * (1 - 0.5 * alpha);
    const coeffPrev1 = 2 * (1 - alpha);
    const coeffPrev2 = (1 - alpha) * (1 - alpha);

    for (let i = 0; i < data.length; i++) {
        if (!Number.isFinite(data[i])) {
            continue;
        }
        if (
            i >= 3
            && Number.isFinite(data[i - 1])
            && Number.isFinite(data[i - 2])
            && Number.isFinite(data[i - 3])
        ) {
            smooth[i] = (data[i] + 2 * data[i - 1] + 2 * data[i - 2] + data[i - 3]) / 6;
        }

        let cycleMain = NaN;
        if (
            i >= 2
            && Number.isFinite(smooth[i])
            && Number.isFinite(smooth[i - 1])
            && Number.isFinite(smooth[i - 2])
            && Number.isFinite(cycle[i - 1])
            && Number.isFinite(cycle[i - 2])
        ) {
            cycleMain =
                coeffCycle * (smooth[i] - 2 * smooth[i - 1] + smooth[i - 2])
                + coeffPrev1 * cycle[i - 1]
                - coeffPrev2 * cycle[i - 2];
        }

        let cycleFallback = NaN;
        if (i >= 2 && Number.isFinite(data[i - 1]) && Number.isFinite(data[i - 2])) {
            cycleFallback = (data[i] - 2 * data[i - 1] + data[i - 2]) / 4;
        }

        cycle[i] = i < 7 ? cycleFallback : cycleMain;
        if (i >= 1 && Number.isFinite(cycle[i - 1])) {
            trigger[i] = cycle[i - 1];
        }
    }

    return { cycle, trigger };
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
    testData = loadTestData();
});

test('ehlers_simple_cycle_indicator_js matches manual reference', () => {
    const n = 160;
    const source = new Float64Array(n);
    for (let i = 0; i < n; i++) {
        source[i] = (testData.high[i] + testData.low[i]) * 0.5;
    }
    const result = wasm.ehlers_simple_cycle_indicator_js(source, 0.07);
    const expected = manualReference(source, 0.07);
    assertArrayClose(result.cycle, expected.cycle, 1e-12, 'cycle mismatch');
    assertArrayClose(result.trigger, expected.trigger, 1e-12, 'trigger mismatch');
});

test('ehlers_simple_cycle_indicator_batch_js first row matches single', () => {
    const n = 128;
    const source = new Float64Array(n);
    for (let i = 0; i < n; i++) {
        source[i] = (testData.high[i] + testData.low[i]) * 0.5;
    }
    const batch = wasm.ehlers_simple_cycle_indicator_batch_js(source, {
        alpha_range: [0.07, 0.09, 0.02],
    });
    const single = wasm.ehlers_simple_cycle_indicator_js(source, 0.07);
    assert.strictEqual(batch.rows, 2);
    assert.strictEqual(batch.cols, n);
    assertArrayClose(batch.alphas, [0.07, 0.09], 1e-12, 'alpha grid mismatch');
    assertArrayClose(batch.cycle.slice(0, n), single.cycle, 1e-12, 'batch cycle mismatch');
    assertArrayClose(batch.trigger.slice(0, n), single.trigger, 1e-12, 'batch trigger mismatch');
});

test('ehlers_simple_cycle_indicator_into pointer path matches safe API', () => {
    const n = 144;
    const source = new Float64Array(n);
    for (let i = 0; i < n; i++) {
        source[i] = (testData.high[i] + testData.low[i]) * 0.5;
    }
    const safe = wasm.ehlers_simple_cycle_indicator_js(source, 0.07);
    const inPtr = wasm.allocate_f64_array(n);
    const outCyclePtr = wasm.ehlers_simple_cycle_indicator_alloc(n);
    const outTriggerPtr = wasm.ehlers_simple_cycle_indicator_alloc(n);
    try {
        wasm.write_f64_array(inPtr, source);
        wasm.ehlers_simple_cycle_indicator_into(inPtr, outCyclePtr, outTriggerPtr, n, 0.07);
        const gotCycle = wasm.read_f64_array(outCyclePtr, n);
        const gotTrigger = wasm.read_f64_array(outTriggerPtr, n);
        assertArrayClose(gotCycle, safe.cycle, 1e-12, 'pointer cycle mismatch');
        assertArrayClose(gotTrigger, safe.trigger, 1e-12, 'pointer trigger mismatch');
    } finally {
        wasm.deallocate_f64_array(inPtr);
        wasm.ehlers_simple_cycle_indicator_free(outCyclePtr, n);
        wasm.ehlers_simple_cycle_indicator_free(outTriggerPtr, n);
    }
});

test('ehlers_simple_cycle_indicator_batch_into returns row count', () => {
    const n = 128;
    const rows = 2;
    const source = new Float64Array(n);
    for (let i = 0; i < n; i++) {
        source[i] = (testData.high[i] + testData.low[i]) * 0.5;
    }
    const inPtr = wasm.allocate_f64_array(n);
    const outCyclePtr = wasm.ehlers_simple_cycle_indicator_alloc(rows * n);
    const outTriggerPtr = wasm.ehlers_simple_cycle_indicator_alloc(rows * n);
    try {
        wasm.write_f64_array(inPtr, source);
        const returnedRows = wasm.ehlers_simple_cycle_indicator_batch_into(
            inPtr,
            outCyclePtr,
            outTriggerPtr,
            n,
            0.07,
            0.09,
            0.02,
        );
        assert.strictEqual(returnedRows, rows);
    } finally {
        wasm.deallocate_f64_array(inPtr);
        wasm.ehlers_simple_cycle_indicator_free(outCyclePtr, rows * n);
        wasm.ehlers_simple_cycle_indicator_free(outTriggerPtr, rows * n);
    }
});

test('ehlers_simple_cycle_indicator_js rejects invalid alpha', () => {
    const source = new Float64Array(testData.close.slice(0, 64));
    assert.throws(
        () => wasm.ehlers_simple_cycle_indicator_js(source, 1.5),
        /Invalid alpha/,
    );
});

File: tests/python/test_ehlers_simple_cycle_indicator.py
import numpy as np
import pytest

from test_utils import load_test_data

try:
    import vector_ta as mp
except ImportError:
    try:
        import vector_ta as mp
    except ImportError:
        pytest.skip(
            "Python module not built. Run 'maturin develop --features python' first",
            allow_module_level=True,
        )


def manual_reference(data: np.ndarray, alpha: float = 0.07):
    cycle = np.full(data.shape[0], np.nan, dtype=np.float64)
    trigger = np.full(data.shape[0], np.nan, dtype=np.float64)
    smooth = np.full(data.shape[0], np.nan, dtype=np.float64)
    coeff_cycle = (1.0 - 0.5 * alpha) * (1.0 - 0.5 * alpha)
    coeff_prev1 = 2.0 * (1.0 - alpha)
    coeff_prev2 = (1.0 - alpha) * (1.0 - alpha)

    for i in range(data.shape[0]):
        if not np.isfinite(data[i]):
            continue
        if i >= 3 and np.isfinite(data[i - 1]) and np.isfinite(data[i - 2]) and np.isfinite(data[i - 3]):
            smooth[i] = (data[i] + 2.0 * data[i - 1] + 2.0 * data[i - 2] + data[i - 3]) / 6.0

        cycle_main = np.nan
        if (
            i >= 2
            and np.isfinite(smooth[i])
            and np.isfinite(smooth[i - 1])
            and np.isfinite(smooth[i - 2])
            and np.isfinite(cycle[i - 1])
            and np.isfinite(cycle[i - 2])
        ):
            cycle_main = (
                coeff_cycle * (smooth[i] - 2.0 * smooth[i - 1] + smooth[i - 2])
                + coeff_prev1 * cycle[i - 1]
                - coeff_prev2 * cycle[i - 2]
            )

        cycle_fallback = np.nan
        if i >= 2 and np.isfinite(data[i - 1]) and np.isfinite(data[i - 2]):
            cycle_fallback = (data[i] - 2.0 * data[i - 1] + data[i - 2]) / 4.0

        cycle[i] = cycle_fallback if i < 7 else cycle_main
        if i >= 1 and np.isfinite(cycle[i - 1]):
            trigger[i] = cycle[i - 1]

    return cycle, trigger


class TestEhlersSimpleCycleIndicator:
    def test_manual_reference_matches_binding(self):
        data = load_test_data()
        source = ((np.asarray(data["high"][:160]) + np.asarray(data["low"][:160])) * 0.5).astype(
            np.float64
        )
        expected_cycle, expected_trigger = manual_reference(source, 0.07)
        result = mp.ehlers_simple_cycle_indicator(source, alpha=0.07)
        np.testing.assert_allclose(result["cycle"], expected_cycle, rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(
            result["trigger"], expected_trigger, rtol=0.0, atol=1e-12, equal_nan=True
        )

    def test_stream_matches_batch(self):
        data = load_test_data()
        source = ((np.asarray(data["high"][:144]) + np.asarray(data["low"][:144])) * 0.5).astype(
            np.float64
        )
        batch = mp.ehlers_simple_cycle_indicator(source, alpha=0.07)
        stream = mp.EhlersSimpleCycleIndicatorStream(0.07)
        streamed = np.array([stream.update(float(x)) for x in source], dtype=np.float64)
        np.testing.assert_allclose(streamed[:, 0], batch["cycle"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(
            streamed[:, 1], batch["trigger"], rtol=0.0, atol=1e-12, equal_nan=True
        )

    def test_batch_metadata_and_first_row_match_single(self):
        data = load_test_data()
        source = ((np.asarray(data["high"][:128]) + np.asarray(data["low"][:128])) * 0.5).astype(
            np.float64
        )
        result = mp.ehlers_simple_cycle_indicator_batch(
            source,
            alpha_range=(0.07, 0.09, 0.02),
        )
        assert result["rows"] == 2
        assert result["cols"] == source.shape[0]
        np.testing.assert_allclose(result["alphas"], np.array([0.07, 0.09]), rtol=0.0, atol=1e-12)
        single = mp.ehlers_simple_cycle_indicator(source, alpha=0.07)
        np.testing.assert_allclose(
            result["cycle"][0], single["cycle"], rtol=0.0, atol=1e-12, equal_nan=True
        )
        np.testing.assert_allclose(
            result["trigger"][0], single["trigger"], rtol=0.0, atol=1e-12, equal_nan=True
        )

    def test_invalid_alpha(self):
        data = load_test_data()
        source = np.asarray(data["close"][:64], dtype=np.float64)
        with pytest.raises(ValueError, match="Invalid alpha"):
            mp.ehlers_simple_cycle_indicator(source, alpha=1.5)
